docs/src/proposals/big-transactions.md

```diff
@@ -0,0 +1,67 @@
+# Big Transactions
+
+## Problem
+Solana networking packets may not exceed the ipv6 MTU size to ensure fast and
+reliable network transmission of cluster information over UDP. Solana's
+networking stack uses a conservative MTU size of 1280 bytes which after
+accounting for networking headers, leaves 1232 bytes for packet data like
+serialized transactions.
+
+Developers building applications on Solana have had to design their on-chain
+programs in a way that works within the transaction size limit constraint. One
+common work-around is to store temporary state on-chain and process the data in
+later transactions. This is the approach used by the BPF loader program for
+deploying Solana programs.
+
+However, this workaround doesn't work well when users want to compose many
+on-chain programs in a single atomic transaction because with more composition,
+comes more account inputs. Each transaction account input takes up 32 bytes and
+as programs get more complex, especially defi protocols, the number accounts
+needed increases quite quickly. There is no available workaround for increasing
+the number of accounts used in a single transaction, it's capped at less than 30
+accounts after signatures and other metadata is accounted for.
+
+## Proposed Solutions
+
+1) Account input compression
+
+Solana can provide a way to map commonly used addresses to small integer values.
+This mapping can be computed on epoch boundaries by all validators or registered
+on-chain in state managed by a new on-chain program.
```

Comment:
I think this pre-registration thing would be a bit of awkward. Also think about some urgent fix needing another registration and waiting for the next epoch, assuming this is high-level defi, where effectively shortened account keys is kind of mandatory for functioning. Sounds a disaster?

Also, the integer values are precious assets by itself. So, we need auction or correct pricing. Otherwise, people are rushed to squat at them?

How about doing some git-like prefix disambiguate thing?



- Make new tx serialization variant so that all accounts (except system account) can be referenced by variable length prefix for its pubkeys  (system account can't be referenced to avoid squatting-attack at same prefix)
- good: generally simpler design, no pre-registration, no transitions of tx formatting for epoch-boundaries
- good: fortunately, our key space is well distributed by sha256 and ed25519.
- bad: reduced size efficiency: maybe 64bits will be needed for practical reasons.
- bad: given tx can fail due to ambiguous account prefix depending on account state at the time, so tx sim is kind of requisite.
- bad: we're more reliant on BTreeMap, which is kind of hard to optimize.

Reply:
@ryoqun that is a cool idea but I'm wondering how we prevent ambiguous accounts if anyone can create accounts at arbitrary addresses by transferring lamports. Are you thinking of additional constraints on the account disambiguation like must be owned by some program or be a PDA?

Reply:
> system account can't be referenced to avoid squatting-attack at same prefix

@jstarry that's exactly the reason I excluded any system accounts at all (those simply created by transferring arbitrary addresses). :) My observations is that defi doesn't need to reference to system accounts so often unless signature check (in that case, we can't use shorthand notation unless breaking our sigcheck pipelining.. xD)

Reply:
Ah that makes perfect sense, thanks for the explanation. I really like the simplicity of this approach.

Also, 64 bit prefix sounds reasonably high.. that's the equivalent of grinding over 10 base58 characters right? 2^64 (1.8e19) ~ 58^11 (2.5e19). But at some point it may be economically worth it to DoS an on-chain protocol by grinding an ambiguous key using cheap compute in the future with some crazy efficient gpu.

Do you have any ideas around future-proofing this approach @ryoqun? The prefix length could be flexible which would allow protocols to adjust to longer prefixes. We could also have a first come first served model for ambiguous accounts to have a strict ordering.

Reply:
@jstarry well I haven't come up with nice protection. that's another `bad:` point... xD

Hopefully, the dos risk should be really low risk. That's because clients just notice some ambiguity at tx sim from rpc after the attacker indeed mounted the squatting attack. Then, client will add another byte to the prefix (then it'll become 9 byte) and resubmit the tx, then done. attacker must spend 256x efforts. So, this is  the reason I made prefix _variable_ and mentioned `git like` :)

So, after spending sizable power/cost for grinding, attacker can only Dos for a slot or two of duration.